src/metabase/models/user_parameter_value.clj

```diff
@@ -30,46 +30,46 @@
   {:value {:in  mi/json-in
            :out json-out}})
 
-(mu/defn upsert!
-  "Upsert or delete parameter value set by the user."
-  [user-id      :- ms/PositiveInt
-   dashboard-id :- ms/PositiveInt
-   parameter-id :- ms/NonBlankString
-   value]
-  (if value
+(mu/defn batched-upsert!
+  "Delete param with nil value and upsert the rest."
+  [user-id         :- ms/PositiveInt
+   dashboard-id    :- ms/PositiveInt
+   param-id->value :- [:map-of :string :any]]
+  (let [to-delete (filter (comp nil? second) param-id->value)
+        to-upsert (filter (comp some? second) param-id->value)]
     (t2/with-transaction [_conn]
+      (doseq [[param-id value] to-upsert]
         (or (pos? (t2/update! :model/UserParameterValue {:user_id      user-id
                                                          :dashboard_id dashboard-id
-                                                         :parameter_id parameter-id}
+                                                         :parameter_id param-id}
                               {:value value}))
             (t2/insert! :model/UserParameterValue {:user_id      user-id
                                                    :dashboard_id dashboard-id
-                                                   :parameter_id parameter-id
+                                                   :parameter_id param-id
                                                    :value        value})))
-    (t2/delete! :model/UserParameterValue
-                :user_id      user-id
-                :dashboard_id dashboard-id
-                :parameter_id parameter-id)))
+      (when (seq to-delete)
+        (t2/delete! :model/UserParameterValue
```

Comment:
This group delete will reduce many db calls if you view a dashboard without any param value